shhext/api.go

```diff
@@ -0,0 +1,185 @@
+	switch event.Event {
+	case whisper.EventEnvelopeSent:
+		confirmed, ok := api.tracker[event.Hash]
+		// if we are not tracking a message or it ws already confirmed - skip it
+		if !ok || confirmed {
+			return
+		}
+		if api.handler != nil {
+			log.Debug("envelope is sent", "hash", event.Hash, "peer", event.Peer)
+			api.handler(event.Hash)
+			api.tracker[event.Hash] = true
+		}
+	case whisper.EventEnvelopeExpired:
+		if _, ok := api.tracker[event.Hash]; ok {
+			log.Debug("envelope expired", "hash", event.Hash)
+			delete(api.tracker, event.Hash)
+		}
+	}
+}
+
+// Post shamelessly copied from whisper codebase with slight modifications.
+func (api *PublicAPI) Post(ctx context.Context, req whisper.NewMessage) (hash common.Hash, err error) {
+	var (
+		symKeyGiven = len(req.SymKeyID) > 0
+		pubKeyGiven = len(req.PublicKey) > 0
+	)
+
+	// user must specify either a symmetric or an asymmetric key
+	if (symKeyGiven && pubKeyGiven) || (!symKeyGiven && !pubKeyGiven) {
+		return hash, whisper.ErrSymAsym
+	}
+
+	params := &whisper.MessageParams{
+		TTL:      req.TTL,
+		Payload:  req.Payload,
+		Padding:  req.Padding,
+		WorkTime: req.PowTime,
+		PoW:      req.PowTarget,
+		Topic:    req.Topic,
+	}
+
+	// Set key that is used to sign the message
+	if len(req.Sig) > 0 {
+		if params.Src, err = api.w.GetPrivateKey(req.Sig); err != nil {
+			return hash, err
+		}
+	}
+
+	// Set symmetric key that is used to encrypt the message
+	if symKeyGiven {
+		if params.Topic == (whisper.TopicType{}) { // topics are mandatory with symmetric encryption
+			return hash, whisper.ErrNoTopics
+		}
+		if params.KeySym, err = api.w.GetSymKey(req.SymKeyID); err != nil {
+			return hash, err
+		}
+		if !validateDataIntegrity(params.KeySym, aesKeyLength) {
+			return hash, whisper.ErrInvalidSymmetricKey
+		}
+	}
+
+	// Set asymmetric key that is used to encrypt the message
+	if pubKeyGiven {
+		params.Dst = crypto.ToECDSAPub(req.PublicKey)
+		if !whisper.ValidatePublicKey(params.Dst) {
+			return hash, whisper.ErrInvalidPublicKey
+		}
+	}
+
+	// encrypt and sent message
+	whisperMsg, err := whisper.NewSentMessage(params)
+	if err != nil {
+		return hash, err
+	}
+
+	env, err := whisperMsg.Wrap(params)
+	if err != nil {
+		return hash, err
+	}
+
+	// send to specific node (skip PoW check)
+	if len(req.TargetPeer) > 0 {
+		n, err := discover.ParseNode(req.TargetPeer)
+		if err != nil {
+			return hash, fmt.Errorf("failed to parse target peer: %s", err)
+		}
+		err = api.w.SendP2PMessage(n.ID[:], env)
+		if err == nil {
+			api.track(env.Hash())
+			return env.Hash(), nil
+		}
+		return hash, err
+	}
+
+	// ensure that the message PoW meets the node's minimum accepted PoW
+	if req.PowTarget < api.w.MinPow() {
+		return hash, whisper.ErrTooLowPoW
+	}
+	err = api.w.Send(env)
+	if err == nil {
+		api.track(env.Hash())
+		return env.Hash(), nil
+	}
+	return hash, nil
+}
+
+// validateDataIntegrity returns false if the data have the wrong or contains all zeros,
```

Comment:
Happy to see descriptive comments even on private functions. 😃

Reply:
that was a copy paste from whisper codebase, i removed it

Reply:
*sigh* 😅